agent/csi/plugin.go

```diff
@@ -2,43 +2,179 @@ package csi
 
 import (
 	"context"
+	"fmt"
+	"sync"
 
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 
 	"github.com/container-storage-interface/spec/lib/go/csi"
 	"github.com/docker/swarmkit/api"
 )
 
 type NodePluginInterface interface {
 	NodeGetInfo(ctx context.Context) (*api.NodeCSIInfo, error)
+	NodeStageVolume(ctx context.Context, req *api.VolumeAssignment) error
+	NodeUnstageVolume(ctx context.Context, req []*api.VolumeAssignment) error
+	NodePublishVolume(ctx context.Context, req []*api.VolumeAssignment) error
+	NodeUnpublishVolume(ctx context.Context, req []*api.VolumeAssignment) error
+}
+
+type volumePublishStatus struct {
+	// stagingPath is staging path of volume
+	stagingPath string
+
+	// isPublished keeps track if the volume is published.
+	isPublished bool
+
+	// publishedPath is published path of volume
+	publishedPath string
 }
 
 // plugin represents an individual CSI node plugin
 type NodePlugin struct {
-	// Name is the name of the plugin, which is also the name used as the
+	// name is the name of the plugin, which is also the name used as the
 	// Driver.Name field
-	Name string
+	name string
+
+	// node ID is identifier for the node.
+	nodeID string
+
+	// socket is the unix socket to connect to this plugin at.
+	socket string
 
-	// Node ID is identifier for the node.
-	NodeID string
+	// cc is the grpc client connection
+	cc *grpc.ClientConn
 
-	// Socket is the unix socket to connect to this plugin at.
-	Socket string
+	// idClient is the identity service client
+	idClient csi.IdentityClient
 
-	// CC is the grpc client connection
-	CC *grpc.ClientConn
+	// nodeClient is the node service client
+	nodeClient csi.NodeClient
 
-	// IDClient is the identity service client
-	IDClient csi.IdentityClient
+	// volumeMap is the map from volume ID to Volume. Will place a volume once it is staged,
+	// remove it from the map for unstage.
+	// TODO: Make this map persistent if the swarm node goes down
+	volumeMap map[string]*volumePublishStatus
+
+	// mu for volumeMap
+	mu sync.RWMutex
+}
+
+const TargetStagePath string = "/var/lib/docker/stage/%s"
+
+const TargetPublishPath string = "/var/lib/docker/publish/%s"
+
+func NewNodePlugin(name string, nodeID string) *NodePlugin {
+	return &NodePlugin{
+		name:      name,
+		nodeID:    nodeID,
+		volumeMap: make(map[string]*volumePublishStatus),
+	}
+}
 
-	// NodeClient is the node service client
-	NodeClient csi.NodeClient
+// GetPublishedPath returns the path at which the provided volume ID is published.
+// Returns an empty string if the volume does not exist.
+func (np *NodePlugin) GetPublishedPath(volumeID string) (string, error) {
+	if volInfo, ok := np.volumeMap[volumeID]; ok {
+		if volInfo.isPublished {
+			return volInfo.publishedPath, nil
+		}
+	}
+	return "", nil
 }
 
 func (np *NodePlugin) NodeGetInfo(ctx context.Context) (*api.NodeCSIInfo, error) {
+	np.mu.RLock()
+	defer np.mu.RUnlock()
```

Comment:
Still need to explain why this lock is acquired here. There is no access to the volumes map, which is what the lock guards.